server/src/main/java/org/elasticsearch/index/mapper/XContentDataHelper.java

```diff
@@ -74,6 +84,58 @@ static void decodeAndWrite(XContentBuilder b, BytesRef r) throws IOException {
         }
     }
 
+    /**
+     * Stores the current parser structure (subtree) to an XContentBuilder and returns it, along with a DocumentParserContext
+     * wrapping it that can be used to reparse the subtree.
+     * The parser of the original context is also advanced to the end of the current structure (subtree) as a side effect.
+     */
+    static Tuple<DocumentParserContext, XContentBuilder> cloneSubContext(DocumentParserContext context) throws IOException {
+        var tuple = cloneSubContextParserConfiguration(context);
+        return Tuple.tuple(cloneDocumentParserContext(context, tuple.v1(), tuple.v2()), tuple.v2());
+    }
+
+    /**
+     * Initializes a XContentParser with the current parser structure (subtree) and returns it, along with a DocumentParserContext
+     * wrapping the subtree that can be used to reparse it.
+     * The parser of the original context is also advanced to the end of the current structure (subtree) as a side effect.
+     */
+    static Tuple<DocumentParserContext, XContentParser> cloneSubContextWithParser(DocumentParserContext context) throws IOException {
+        var tuple = cloneSubContextParserConfiguration(context);
+        XContentParser parser = XContentHelper.createParserNotCompressed(
+            tuple.v1(),
+            BytesReference.bytes(tuple.v2()),
+            context.parser().contentType()
+        );
+        assert parser.currentToken() == null;
+        parser.nextToken();
+        return Tuple.tuple(cloneDocumentParserContext(context, tuple.v1(), tuple.v2()), parser);
+    }
+
+    private static Tuple<XContentParserConfiguration, XContentBuilder> cloneSubContextParserConfiguration(DocumentParserContext context)
+        throws IOException {
+        XContentParser parser = context.parser();
+        XContentBuilder builder = XContentBuilder.builder(parser.contentType().xContent());
+        builder.copyCurrentStructure(parser);
+
+        XContentParserConfiguration configuration = XContentParserConfiguration.EMPTY.withRegistry(parser.getXContentRegistry())
+            .withDeprecationHandler(parser.getDeprecationHandler())
+            .withRestApiVersion(parser.getRestApiVersion());
+        return Tuple.tuple(configuration, builder);
+    }
+
+    private static DocumentParserContext cloneDocumentParserContext(
+        DocumentParserContext context,
+        XContentParserConfiguration configuration,
+        XContentBuilder builder
+    ) throws IOException {
+        DocumentParserContext subcontext = context.switchParser(
```

Comment:
Done.

Reply:
Looking at the concerns we had with switchParser, i think your assessment is right. Also, its javadocs says

>  we are actively deprecating and removing the ability to pass complex objects to multifields, so try and avoid using this method

Given that this was committed 4 years ago, the "actively" part no longer holds true. I reviewed the current usages of switchParser and they seem ok to me. I have a slight concern that we are now using the method in a general utility method, that may spread its usage instead of limiting it to the only places where it's strictly needed. I am not sure what to do about that though, and I am ok with the current plan. Thanks for raising this.